packages/search-ui/src/DebounceManager.js

```diff
@@ -1,30 +1,58 @@
 import debounceFn from "debounce-fn";
 
-export default class DebounceManager {
+class DebounceManager {
   debounceCache = {};
 
-  /*
-  The purpose of this is to:
-  Dynamically debounce and cache a debounced version of a function at the time of calling that function. This avoids
-  managing debounced version of functions locally.
-
-  Assumption:
-  Functions are debounced on a combination of unique function and wait times. So debouncing won't work on
-  subsequent calls with different wait times or different functions. That also means that the debounce manager
-  can be used for different functions in parallel, and keep the two functions debounced separately.
-
-  */
-  runWithDebounce(wait, fn, ...parameters) {
+  /**
+   * Dynamically debounce and cache a debounced version of a function at the time of calling that function. This avoids
+   * managing debounced version of functions locally.
+   *
+   * In other words, debounce usually works by debouncing based on
+   * referential identity of a function. This works by comparing provided function names.
+   *
+   * This also has the ability to short-circuit a debounce all-together, if no wait
+   * time is provided.
+   *
+   * Assumption:
+   * Functions are debounced on a combination of unique function name and wait times. So debouncing won't work on
+   * subsequent calls with different wait times or different functions. That also means that the debounce manager
+   * can be used for different functions in parallel, and keep the two functions debounced separately.
+   *
+   *
+   * @param {number} wait Milliseconds to debounce. Executes immediately if falsey.
+   * @param {function} fn Function to debounce
+   * @param {function} functionName Name of function to debounce, used to create a unique key
+   * @param  {...any} parameters Parameters to pass to function
+   */
+  runWithDebounce(wait, functionName, fn, ...parameters) {
     if (!wait) {
       return fn(...parameters);
     }
 
-    const key = fn.toString() + wait.toString();
+    const key = functionName + wait.toString();
     let debounced = this.debounceCache[key];
     if (!debounced) {
       this.debounceCache[key] = debounceFn(fn, { wait });
       debounced = this.debounceCache[key];
     }
     debounced(...parameters);
   }
+
+  cancelByName(functionName) {
+    Object.entries(this.debounceCache)
+      .filter(([cachedKey]) => cachedKey.startsWith(functionName))
```

Comment:
I'm a little leery here of using `startsWith` instead of just `cachedKey === functionName`. It would return a false positive for `foo()` vs. `fooBar()`, however unlikely. Is there a particular reason to use `startsWith`?

Reply:
That's a great point! The functions are cached by function name and time. So we could have:

```
{
  someFunction1000: {},
  someFunction500: {},
  someFunction100: {},
}
```

I could simply change it to be:
```
{
  someFunction|1000: {},
  someFunction|500: {},
  someFunction|100: {},
}
```

And do `startsWith(`${functionName}|`).

Reply:
Ahh gotcha! My bad, I hadn't realized we were combining the time with the name for the cache key (although that makes a lot of sense ha).

I think `|` should work really well as a separator since JS functions [can't include that character](https://stackoverflow.com/questions/1661197/what-characters-are-valid-for-javascript-variable-names) in a valid function name! 👍